aws-sigv4/src/types.rs

```diff
@@ -50,16 +60,50 @@ impl CanonicalRequest {
             creq.params = qs::to_string(params)?;
         }
 
-        let mut headers = BTreeSet::new();
-        for (name, _) in req.headers() {
-            headers.insert(CanonicalHeaderName(name.clone()));
+        let mut canonical_headers = req.headers().clone();
+        canonical_headers.remove("x-amz-date");
+        canonical_headers.remove("x-amz-security-token");
+        let x_amz_date = HeaderName::from_static("x-amz-date");
```

Comment:
All this newly-added code is... a lot. What's it being used for? Can the different kinds of canonical requests for different services (S3, EC2, REST-JSON-based services, etc.) be documented in doc comments?

Reply:
1. we need to add `x-amz-date` and `x-security-token` in the canonical request. This really should apply to all services, they just happen to accept it
2. some services (s3, s3 control) also require a checksum header added
3. streaming bodies need to be able to be "unsigned payload"
Will add more docs for all of this.